src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,155 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> firstNode;
+    private Node<T> lastNode;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (firstNode == null) {
+            Node<T> newNode = new Node<>(null, value, null);
+            firstNode = newNode;
+            lastNode = newNode;
+        } else {
+            lastNode.next = new Node<>(lastNode, value, null);
+            lastNode = lastNode.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, firstNode);
+            if (firstNode != null) {
+                firstNode.prev = newNode;
+            }
+            firstNode = newNode;
+            if (lastNode == null) {
+                lastNode = newNode;
+            }
+        } else {
+            Node<T> current = firstNode;
+            for (int i = 0; i < index - 1; i++) {
+                current = current.next;
+            }
+            Node<T> nextNode = current.next;
+            Node<T> newNode = new Node<>(current, value, nextNode);
+            current.next = newNode;
+            if (nextNode != null) {
+                nextNode.prev = newNode;
+            } else {
+                lastNode = newNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
```

Comment:
The `addAll(List<T> list)` method does not handle `null` values in the list correctly. Consider adding checks for `null` values within the list.

Reply:
IIRC original implementation doesn't check this
empty list just won't be iterated with for each, and null will throw NPE